### ISSUE TITLE
torch random number generation is inconsistent across GPUs

### ISSUE THREAD

**Issue by antoche** [pytorch/pytorch#62467]
## 🐛 Bug

Running the exact same random number generation path across two different devices, even ensuring deterministic behaviour, gives me different results.

## To Reproduce

```python
#!/usr/bin/env python

import os
import torch

# Force determinism
torch.backends.cudnn.benchmark = False
torch.backends.cudnn.deterministic = True
os.putenv("CUBLAS_WORKSPACE_CONFIG", ":4096:8")
torch.set_deterministic(True)
torch.use_deterministic_algorithms(True)
# Force full-precision
torch.backends.cuda.matmul.allow_tf32 = False
torch.backends.cudnn.allow_tf32 = False

print('Device 0: {}'.format(torch.cuda.get_device_name(0)))
print('Device 1: {}'.format(torch.cuda.get_device_name(1)))

rng0 = torch.Generator(device='cuda:0').manual_seed(43)
rng1 = torch.Generator(device='cuda:1').manual_seed(43)

shape = (1, 3, 128, 128)
noise0 = torch.empty(*shape, device='cuda:0').normal_(generator=rng0)
noise1 = torch.empty(*shape, device='cuda:1').normal_(generator=rng1)

if not torch.allclose(noise0.cpu(), noise1.cpu(), atol=0., rtol=0.):
    print(noise0[...,0,0])
    print(noise1[...,0,0])
```

Running the script above results in this:

```
Device 0: RTX A6000
Device 1: Quadro M4000
tensor([[1.9983, 0.8704, 2.1127]], device='cuda:0')
tensor([[1.9983, 0.8704, 0.1787]], device='cuda:1')
```

Even though some values are identical (e.g., the first two channels in this case), not all of them are.

## Expected behavior

As mentioned in this [forum post](https://discuss.pytorch.org/t/manual-seed-non-deterministic-on-different-gpus/44549/2), I should see consistent results even running on different physical GPUs. Nothing should be printed other than the device names, as both `noise0` and `noise1` ought to be the same.

## Environment

 - PyTorch Version: 1.8.0
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (conda, pip, source): source
 - Python version: 3.7.3
 - CUDA/cuDNN version: 11.1
 - GPU models and configuration: Nvida RTX A6000 + Nvidia Quadro M4000

cc @ngimel @pbelevich @heitorschueroff

**Comment by ngimel** [comment]
This is expected, pytorch does not guarantee reproducibility across different devices https://pytorch.org/docs/stable/notes/randomness.html?highlight=deterministic.

**Comment by antoche** [comment]
So @colesbury was incorrect when they wrote "You should get consistent random numbers if you’re using the same seed, PyTorch version, and CUDA version even if it’s run on a different physical GPU" ?

**Comment by ngimel** [comment]
It may have been correct at the time, but currently number of blocks launched by rng kernels depends on the number of SMs in the GPU, so GPUs with the different number of SMs produce different results.

**Comment by antoche** [comment]
I see, thanks for the explanation.
